docs/spec_draft.md

```diff
@@ -522,6 +523,83 @@ it only exists to establish data dependencies from `result` to `inputs`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups).
+
+Afterwards, within each `process_group`:
```

Comment:
Result for every process is computed as outlined in the following Python-like syntax:
```python
for process_group in process_groups:
  for i in process_group:
    result[i0, i1, ..., iR-1]@process[i] = reduce(
      inputs = [ operand[i0, i1, ..., iR-1]@process[j] for j in process_group ],
      init_values = [x], # such that `computation` and `x` form a monoid
      dimensions = [0],
      body = computation
    )
```

Reply:
We can omit the sender and receiver part.

Reply:
I have adopted peices of the suggestions. 

Overall,
The sender receiver part is important to convey the communication aspect of the op: That is, the a receiver receives all the  operands from the senders before reducing them locally. 
Note that the second bullet does not use any sender/receiver just to highlight that the computation is local.